src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -499,19 +502,87 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 return Ok(res.try_into().unwrap());
             }
             Err(ResolutionFailure::NotResolved {
+                item_id,
                 module_id,
                 partial_res: None,
                 unresolved: path_str.into(),
             })
         })
     }
 
+    fn resolve_self_ty(&self, path_str: &str, ns: Namespace, item_id: ItemId) -> Option<Res> {
+        if ns != TypeNS || path_str != "Self" {
+            return None;
+        }
+
+        let self_id = match item_id.as_def_id() {
+            None => None,
+            Some(did)
+                if (matches!(self.cx.tcx.def_kind(did), DefKind::Field)
+                    && matches!(
+                        self.cx.tcx.def_kind(self.cx.tcx.parent(did).unwrap()),
+                        DefKind::Variant
+                    )) =>
+            {
+                self.cx.tcx.parent(did).and_then(|item_id| self.cx.tcx.parent(item_id))
+            }
+            Some(did)
+                if matches!(
+                    self.cx.tcx.def_kind(did),
+                    DefKind::AssocConst
+                        | DefKind::AssocFn
+                        | DefKind::AssocTy
+                        | DefKind::Variant
+                        | DefKind::Field
+                ) =>
+            {
+                self.cx.tcx.parent(did)
+            }
+            Some(did) => Some(did),
+        };
+
+        self_id.map(|self_id| {
+            let def_kind = self.cx.tcx.def_kind(self_id);
+            match def_kind {
+                DefKind::Impl => match self.cx.tcx.type_of(self_id).kind() {
+                    ty::Adt(adt_def, _) => {
+                        let kind = match adt_def.adt_kind() {
+                            AdtKind::Struct => DefKind::Struct,
+                            AdtKind::Union => DefKind::Union,
+                            AdtKind::Enum => DefKind::Enum,
+                        };
+                        Res::Def(kind, adt_def.did)
+                    }
+                    ty::Foreign(def_id) => Res::Def(DefKind::ForeignTy, *def_id),
+                    ty::Int(int_ty) => Res::Primitive((*int_ty).into()),
+                    ty::Uint(uint_ty) => Res::Primitive((*uint_ty).into()),
+                    ty::Float(float_ty) => Res::Primitive((*float_ty).into()),
+                    ty::Bool => Res::Primitive(PrimitiveType::Bool),
+                    ty::Char => Res::Primitive(PrimitiveType::Char),
+                    ty::Str => Res::Primitive(PrimitiveType::Str),
+                    _ => Res::Def(def_kind, self_id),
```

Comment:
What about all the other primitives, for example slices?

Reply:
This defkind is wrong as well.

Reply:
Majority of the code in this file is wrong to some extent, to be honest.
So my additions keep their hackiness roughly on the same level with the surrounding code, otherwise you'll have to rewrite everything here to do things correctly.

Here I only need to consider inherent impls that are later considered by `resolve_primitive_associated_item`.
And that's all the cases listed in rustdoc's `enum PrimitiveType` as I understand?
And more specifically, the variants used in `fn all_impls`.
In that case I need to extend this list, yes.

Reply:
Done.
I've reused `fn def_id_to_res` here.